File: app/src/main/java/com/peterlaurence/trekadvisor/menu/mapview/components/PathView.java
package com.peterlaurence.trekadvisor.menu.mapview.components;


import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.View;

import java.util.Arrays;
import java.util.HashSet;

/**
 * This is a modified version of {@link com.qozix.tileview.paths.CompositePathView}, using
 * {@code Canvas.drawLines} to draw a path. This method is much more efficient as it's
 * hardware accelerated, although the result is not as neat as the original implementation (which
 * uses {@link Path}.
 *
 * @author peterLaurence on 19/02/17
 */
public class PathView extends View {

    private static final int DEFAULT_STROKE_COLOR = 0xCC311B92;
    private static final int DEFAULT_STROKE_WIDTH_DP = 4;
    private float mStrokeWidthDefault;

    private float mScale = 1;

    private boolean mShouldDraw = true;

    private HashSet<DrawablePath> mDrawablePaths = new HashSet<>();

    private Paint mDefaultPaint = new Paint();


    public PathView(Context context) {
        super(context);
        setWillNotDraw(false);

        DisplayMetrics metrics = getResources().getDisplayMetrics();
        mStrokeWidthDefault = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, DEFAULT_STROKE_WIDTH_DP, metrics);

        mDefaultPaint.setStyle(Paint.Style.STROKE);
        mDefaultPaint.setColor(DEFAULT_STROKE_COLOR);
        mDefaultPaint.setStrokeWidth(mStrokeWidthDefault);
        mDefaultPaint.setAntiAlias(true);
        mDefaultPaint.setStrokeJoin(Paint.Join.ROUND);
        mDefaultPaint.setStrokeCap(Paint.Cap.ROUND);
    }

    public float getScale() {
        return mScale;
    }

    public void setScale(float scale) {
        mScale = scale;
        invalidate();
    }

    public Paint getDefaultPaint() {
        return mDefaultPaint;
    }

    public void addPath(DrawablePath drawablePath) {
        if (drawablePath.paint == null) {
            drawablePath.paint = mDefaultPaint;
            drawablePath.width = mStrokeWidthDefault;
        }

        mDrawablePaths.add(drawablePath);
        invalidate();
    }

    public void removePath(DrawablePath path) {
        mDrawablePaths.remove(path);
        invalidate();
    }

    public void clear() {
        mDrawablePaths.clear();
        invalidate();
    }

    public void setShouldDraw(boolean shouldDraw) {
        mShouldDraw = shouldDraw;
        invalidate();
    }

    @Override
    public void onDraw(Canvas canvas) {
        if (mShouldDraw) {
            canvas.scale(mScale, mScale);
            for (DrawablePath drawablePath : mDrawablePaths) {
                if (drawablePath.visible) {
                    drawablePath.paint.setStrokeWidth(drawablePath.width / mScale);
                    canvas.drawLines(drawablePath.path, drawablePath.paint);
                }
            }
        }
        super.onDraw(canvas);
    }

    public static class DrawablePath {

        public DrawablePath(float[] path, Paint paint) {
            this.path = path;
            this.paint = paint;
            visible = true;
        }

        @Override
        public int hashCode() {
            return Arrays.hashCode(path);
        }

        /**
         * The path that this drawable will follow.
         */
        public float[] path;

        /**
         * The paint to be used for this path.
         */
        public Paint paint;

        /**
         * Whether or not this path shall be drawn
         */
        public boolean visible;

        /**
         * The width of the path
         */
        public float width;
    }
}

File: app/src/main/java/com/peterlaurence/mapview/paths/PathView.kt
package com.peterlaurence.mapview.paths


import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.util.TypedValue
import android.view.View
import com.peterlaurence.mapview.MapView
import com.peterlaurence.mapview.ScaleChangeListener

/**
 * This is a custom view that uses, using [Canvas.drawLines] to draw a path.
 * This method is much more efficient as it's hardware accelerated, although the result is not as
 * neat as the output of [Path].
 *
 * @author peterLaurence on 19/02/17 -- Converted to Kotlin on 26/07/19
 */
class PathView(context: Context) : View(context), ScaleChangeListener {
    private val strokeWidthDefault: Float

    var scale = 1f
        set(scale) {
            field = scale
            invalidate()
        }

    private var shouldDraw = true

    private var pathList: List<DrawablePath>? = null

    private val defaultPaint = Paint()

    /**
     * The color of the default [Paint] that is assigned to a [DrawablePath] if its [DrawablePath.paint]
     * property isn't set.
     */
    var color: Int
        get() = defaultPaint.color
        set(value) {
            defaultPaint.color = value
        }

    init {
        setWillNotDraw(false)

        val metrics = resources.displayMetrics
        strokeWidthDefault = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, DEFAULT_STROKE_WIDTH_DP.toFloat(), metrics)

        defaultPaint.style = Paint.Style.STROKE
        defaultPaint.color = DEFAULT_STROKE_COLOR
        defaultPaint.strokeWidth = strokeWidthDefault
        // As of 22/06/19, the settings below cause performance drop when long Routes are rendered
        //        mDefaultPaint.setAntiAlias(true);
        //        mDefaultPaint.setStrokeJoin(Paint.Join.ROUND);
        //        mDefaultPaint.setStrokeCap(Paint.Cap.ROUND);
    }

    fun updatePaths(pathList: List<DrawablePath>) {
        this.pathList = pathList
        invalidate()
    }

    override fun onScaleChanged(scale: Float) {
        this.scale = scale
    }

    fun setShouldDraw(shouldDraw: Boolean) {
        this.shouldDraw = shouldDraw
        invalidate()
    }

    public override fun onDraw(canvas: Canvas) {
        val pathList = this.pathList
        if (shouldDraw && pathList != null) {
            canvas.scale(scale, scale)
            for (path in pathList) {
                /* If no Paint is defined, take the default one */
                if (path.paint == null) {
                    path.paint = defaultPaint
                }

                /* If no width is defined, take the default one */
                val width = path.width ?: strokeWidthDefault

                if (path.visible) {
                    path.paint?.let {
                        it.strokeWidth = width / scale
                        canvas.drawLines(path.path, it)
                    }
                }
            }
        }
        super.onDraw(canvas)
    }

    interface DrawablePath {
        /**
         * Whether or not this path should be drawn
         */
        val visible: Boolean
        /**
         * The path. See [Canvas.drawLines].
         */
        var path: FloatArray
        /**
         * The paint to be used for this path.
         */
        var paint: Paint?
        /**
         * The width of the path
         */
        val width: Float?
    }
}

private const val DEFAULT_STROKE_COLOR = -0xc0ae4b
private const val DEFAULT_STROKE_WIDTH_DP = 4

/**
 * Helper function to correctly add a [PathView] to the [MapView].
 */
fun MapView.addPathView(pathView: PathView) {
    addView(pathView, 1)
    addScaleChangeListener(pathView)
}

/**
 * Helper function to correctly remove a [PathView] from the [MapView].
 */
fun MapView.removePathView(pathView: PathView) {
    removeView(pathView)
    removeScaleChangeListener(pathView)
}